src/lib.rs

```diff
@@ -36,6 +36,279 @@
 )]
 // This one is hard to avoid.
 #![allow(clippy::multiple_crate_versions)]
+//xx For now:
+#![allow(clippy::not_unsafe_ptr_arg_deref)]
+#![allow(missing_debug_implementations)]
+#![allow(missing_docs)]
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    dead_code,
+    trivial_casts
+)]
+#[allow(clippy::all)]
+mod psa_crypto_binding {
+    include!(concat!(env!("OUT_DIR"), "/shim_bindings.rs"));
+}
+
+#[allow(dead_code)]
+mod constants;
+pub use constants::*;
+
+struct Global {
+    init_succeeded: bool,
+    // Calls to psa_open_key, psa_generate_key and psa_destroy_key
+    // are not thread safe. We work around this bug with the key_mutex.
+    // Mbed issue: https://github.com/ARMmbed/mbed-crypto/issues/266
+    key_mutex: Mutex<()>,
+}
+
+impl Global {
+    fn new() -> Global {
+        Global {
+            init_succeeded: false,
+            key_mutex: Mutex::new(()),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLOBAL: Mutex<Global> = Mutex::new(Global::new());
+}
+
+fn init() -> bool {
+    if GLOBAL.lock().unwrap().init_succeeded {
+        return true;
+    }
+    let status = unsafe { psa_crypto_binding::psa_crypto_init() };
+    if status != PSA_SUCCESS {
+        return false;
+    }
+    GLOBAL.lock().unwrap().init_succeeded = true;
+    true
+}
+
+macro_rules! wrap_any {
+    ($x:expr) => {
+        if !init() {
+            panic!("Error when initialising PSA Crypto")
+        } else {
+            #[allow(unused_unsafe)]
+            unsafe {
+                $x
+            }
+        }
+    };
+}
+
+macro_rules! wrap_status {
+    ($x:expr) => {
+        if !init() {
+            PSA_ERROR_BAD_STATE
+        } else {
+            unsafe { $x }
+        }
+    };
+}
+
+macro_rules! key_lock {
+    ($x:expr) => {{
+        let mutex = &GLOBAL.lock().unwrap().key_mutex;
+        let _guard = mutex.lock().unwrap();
+        $x
+    }};
+}
+
+// Reexported types:
+
+pub use psa_crypto_binding::psa_algorithm_t;
+pub use psa_crypto_binding::psa_key_handle_t;
+pub use psa_crypto_binding::psa_key_id_t;
+pub use psa_crypto_binding::psa_key_lifetime_t;
+pub use psa_crypto_binding::psa_key_type_t;
+pub use psa_crypto_binding::psa_key_usage_t;
+pub use psa_crypto_binding::psa_status_t;
+
+// Wrapped types:
+
+#[allow(non_camel_case_types)]
+pub struct psa_key_attributes_t {
+    x: psa_crypto_binding::psa_key_attributes_t,
+}
```

Comment:
This is ok as is for now, but we'll have to decide if we want to keep the naming convention from the C spec or if we want to make this truly Rust-native and use the camel case naming convention.

Also, if you're not interested in having a proper name for the internal value, you can simply do
```Rust
pub struct psa_key_attributes_t(psa_crypto_binding::psa_key_attributes_t);
```
and access it through `self.0`

Reply:
Converting names to the Rust naming convention seems like a good idea to me. Anyone else in favour or opposed?

Reply:
I think it is a good idea too but maybe in a later PR where we "rustify" other things as well?